### ISSUE TITLE
auto-formatting all code

### ISSUE THREAD

**Issue by cos** [kframework/k-legacy#1852]
Skimming through formatting-only changes when reviewing pull requests can be annoying. On the other hand, being careful to always maintain formatting manually is also a pain. It is much easier to tell the IDE to auto-format the entire file after making some changes than to go line by line to make sure everything is correctly formatted. Furthermore, the manual process is bound to include formatting mistakes -- even as simple as forgetting to select some line when auto-formatting only a selection.

I propose we never format our files manually. Yes, there might be 1 in 100 cases where the manual formatting will look prettier, but I strongly believe it is not worth inconveniencing the other 99 cases.

Please let me know if you ok with this. 

If we do agree on this, we have to figure out a way to transition without too much disruption. The main issue are the pending PRs. I think the following should work: I will create a new branch off master which has everything auto-formatted. Everyone who has a significant PR open could auto-format locally (please make sure to use our [committed IntelliJ style file](https://github.com/kframework/k/blob/master/.idea/codeStyleSettings.xml)) and then try to merge with the new master. If it works without too much merge pain, everything's great. If it doesn't, let me know so we can try to figure out a different solution.

If everything works well, I'll merge the auto-format branch in master.

If I don't hear from people by the end of the work week I will assume that: (1) they agree with the change and (2) they don't have any significant issues with their pull requests.

Pinging people with recent PRs: @msaxena2, @yilongli, @daejunpark, @edgar-pek, @radumereuta, @andreistefanescu 

And people who might have something to say on the subject: @dwightguth, @bmmoore, @grosu

**Comment by dwightguth** [comment]
I don't have any particularly strong objection to this if it actually works, but I am pretty sure I have seen cases in the past where stuff formatted with the same style file on different computers generated different results, possibly due to different versions of IntelliJ or something. I don't really know for sure. This also presumes that everyone will edit only with IntelliJ which I'm not sure we can assume either, so if you're going to do this, you're going to need to come up with some way to check the style at build time as well.

**Comment by bmmoore** [comment]
I definitely agree we should just settle on some automatic formatting, but we should also keep formatting changes separate from code changes. If there were formatting errors, you should autoformat and commit before you start changing stuff. But that shouldn't matter once we can fix formatting and keep it fixed.

One big commit to fix formatting plus infrastructure to keep it fixed sounds good.

Can we run formatting as a script without starting the IntelliJ interface, and have formatting checked automatically?

**Comment by cos** [comment]
@dwightguth, yes, I have seen inconsistencies between IntelliJ versions too. Still, I'm not aiming for perfection here. If we can eliminate 95% of PR junk without having too worry about formatting while we code, I'm more than happy.

@bmmoore, checking formatting automatically (maybe as a github hook) would be nice, but it will require a bit more work. We could look into it once we have things formatted.

**Comment by cos** [comment]
Here is a pull request with the result of the auto-format:https://github.com/kframework/k/pull/1853 

Please cherry-pick the IntelliJ style file separately when testing this against your PRs -- would it help if I make a pull request against master for this separately?

Have a look over the result and feel free to make adjustments to the style file as you see fit. I've added a couple of options that partially address one thing (extra new lines) @yilongli didn't like before.

**Comment by cos** [comment]
Also, feel free to change that branch. Please force-push your auto-format result so we add less junk to the history.

**Comment by yilongli** [comment]
I do not agree that we should never do manual formatting. I may choose to do manual formatting when the auto-formatting looks bad and I don't want my pull request to be rejected because of not using auto-formatting.

Regarding pending pull requests, PLEASE do not make merging my pull requests any harder. I have finished those pull requests weeks ago. I only agree to postpone them because @andreistefanescu wants to merge the automaton PR first.

**Comment by cos** [comment]
@yilongli, do you have an example where automatic formatting looks unacceptably bad?

Regarding pending pull requests, please try the process I've suggested and let me know if you encounter signifiant issues. If you do, we'll figure out a way around it or, worst case, will postpone the auto-format for after you merge.

**Comment by yilongli** [comment]
@cos I do not have a concrete example at hand but I want certain degree of flexibility.

I have three pending pull requests instead of one. I don't think there will be significant issues but I don't want to spend any more time on them. Could you make merging that automaton PR your first priority?

**Comment by cos** [comment]
> @cos I do not have a concrete example at hand but I want certain degree of flexibility.

@yilongli, well, as we won't likely have an automatic check, your PR will probably go through, but the custom formatting may get removed later on when someone else auto-formats. Examples of ugly automatic formatting would be useful as we might be able to convince IntelliJ to format it differently.

> I have three pending pull requests instead of one. I don't think there will be significant issues but I don't want to spend any more time on them. Could you make merging that automaton PR your first priority?

Hopefully today.

**Comment by yilongli** [comment]
@cos OK then, I can accept that we may revise the IntelliJ style file when we see ugly automatic formatting.

**Comment by radumereuta** [comment]
@cos can you give us some examples of big projects that use this "enforced" auto-formatting? What we have right now is inspired from what Google does, if I remember correctly.
@yilongli I found an example where this fails badly: https://github.com/kframework/k/blob/auto-format-everything/kernel/src/main/java/org/kframework/parser/concrete2kore/kernel/Grammar.java

In the beginning comments I used ASCII to describe how the internals of the parser works.
Auto-formatting messes up the alignment making it unreadable.

**Comment by cos** [comment]
> @cos can you give us some examples of big projects that use this "enforced" auto-formatting? What we have right now is inspired from what Google does, if I remember correctly.

I think many developers do not worry that much about this because either they are less fussy about the noise in pull requests, or have set up automated systems to deal with the noise. In most cases it is simply considered common courtesy to closely follow the project's guidelines and projects help by providing a style file for automatic formatting. So this is not a front and center issue usually. Still, here's a couple of examples:

**jQuery**
The people at jQuery want to, long-term, reach the "nirvana" of code formatting by doing it automatically as part of the build process: https://contribute.jquery.org/style-guide/
https://contribute.jquery.org/style-guide/

**go language**
They advise you to install a git hook to gofmt, an automatic formatter:
https://golang.org/doc/contribute.html#tmp_7

Also, to get a feeling of other developers' opinion:
http://programmers.stackexchange.com/questions/197257/is-it-a-good-idea-to-format-the-code-in-eclipse-using-auto-format
http://programmers.stackexchange.com/questions/189274/is-imposing-the-same-code-format-for-all-developers-a-good-idea

> @yilongli I found an example where this fails badly: https://github.com/kframework/k/blob/auto-format-everything/kernel/src/main/java/org/kframework/parser/concrete2kore/kernel/Grammar.java
> 
> In the beginning comments I used ASCII to describe how the internals of the parser works.
> Auto-formatting messes up the alignment making it unreadable.

Actually, using ASCII art in javadoc is not such a good idea. Here's how your original file gets rendered: http://cosmin.radoi.net/javadoc-test/index.html

In this case, it might be better to either embed an actual image (you can do that in javadoc), or move the documentation to the wiki and link to it.

That being said, if you still really want to use it, we could either disable automatic formatting of javadoc or you could use extra characters to ensure it doesn't break your alignment.

**Comment by radumereuta** [comment]
Ok, I like your answer.
I would prefer solving the javadoc issue from Grammar.java differently. Just make the comment simple multiline comment. Apparently IntelliJ doesn't touch those.

I also propose that we make such a change in steps. Maybe start with Scala files (because we have fewer files), and then after we test such an enforced system, add it for Java, XML, and other types of files. Also, we may want to have exceptions (example may be other copyrighted files).

**Comment by bmmoore** [comment]
Adding a `<pre>` tag to make it display properly in the browser will also stop IntelliJ's autoformatting from touching the alignment.

**Comment by cos** [comment]
@bmmoore, great, didn't know about that. Works well.

@radumereuta, , I added `<pre>` around the figure you mentioned.

**Comment by cos** [comment]
The pull request (#1853) is ready for merging. Please have a look at it beforehand.

**Comment by cos** [comment]
@radumereuta, you have a critical eye for code formatting detail, and have already reviewed a good part of it. Could you review the rest?

**Comment by cos** [comment]
Regarding outstanding pull requests. I tried to merge one of my branches (#1873) into the auto-format branch. It's not perfectly smooth, but doable. Cherry-pick e746016dd0555ddbf55d552416e1ed7aaabd32f6, then autoformat in IntelliJ with *.scala and then with *.java masks. Then try to merge. I got three false merge conflicts but it wasn't that bad -- kind of similar to what we were facing with SVN. 

Let me know if you run into any trouble. We don't have very large outstanding PRs right now so I don't think we'll find a better time to do this.

**Comment by bmmoore** [comment]
For @yilongli, IntelliJ does allow at least some degree of manual control. It normalizes indentation and trailing spaces, but if your lines are not too long it generally leaves linebreaks alone. For example, this code is left untouched (in an appropriate context in I file I was just working on):

```
        s = sortCells.sortCells(s);
        s = sortCells
                .sortCells(s);
```

**Comment by cos** [comment]
hm... I looked over the merge commit for #1873 and there are some problems. The auto-format commit makes formatting changes seem like very new code, possibly overriding your code, or your auto-format overriding newer code in the master.

**Comment by cos** [comment]
Hm... the following should work:
1. Merge the latest master.
2. Cherry-pick e746016dd0555ddbf55d552416e1ed7aaabd32f6
3. Auto-format with *.java and *.scala masks
4. In `Reflection.scala` delete the space before "default" (IntelliJ has a bug when auto-formatting Scala interpolated string.
5. Commit
6. Merge the auto-format-everything branch.

**Comment by cos** [comment]
The above steps gave a clean, fast-forward merge for my `feature/configuration-inheritance` branch.
